### PR TITLE
VS-102 : Replace Type Names in Error Messages

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**
Replace Type Names in Error Messages

### CODE REVIEW

File: src/MongoDB.Analyzer/Core/Builders/BuildersAnalyzer.cs [hunk 1]
```diff
@@ -68,6 +69,9 @@ private static AnalysisStats ReportMqlOrInvalidExpressions(MongoAnalysisContext
         var settings = context.Settings;
         int mqlCount = 0, internalExceptionsCount = 0, driverExceptionsCount = 0;
 
+        var typesMapper = new TypesMapper();
+        typesMapper.AddMappings(MqlGeneratorSyntaxElements.Builders.MqlGeneratorNamespace, context.TypesProcessor);
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Builders/BuildersAnalyzer.cs:73]: Lets move `AddMappings` to `TypesMapper` ctor for simplicity. And create it only when needed:
`typesMapper ?? = new TypesMapper();`.

File: src/MongoDB.Analyzer/Core/TypesProcessor.cs [hunk 2]
```diff
@@ -22,6 +22,8 @@ internal sealed class TypesProcessor
 
     public MemberDeclarationSyntax[] TypesDeclarations => _processedTypes.Values.Select(p => p.NewDeclaration).ToArray();
 
+    public Dictionary<string, string> GeneratedTypeToOriginalTypeMapping => _processedTypes.ToDictionary(x => x.Value.NewName, x => x.Key);
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/TypesProcessor.cs:25]: Turn it to IEnumerable<(NewName, PreviousName)>, to skip Dictionary object creation.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/TypesProcessor.cs (hunk 2 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/TypesProcessor.cs:25]: Done

File: src/MongoDB.Analyzer/Core/Builders/BuildersAnalyzer.cs (hunk 1 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Builders/BuildersAnalyzer.cs]: Done

File: src/MongoDB.Analyzer/Core/Linq/LinqAnalyzer.cs [hunk 3]
```diff
@@ -113,7 +114,9 @@ private static AnalysisStats ReportMqlOrInvalidExpressions(MongoAnalysisContext
                 if (isDriverOrLinqException || settings.OutputInternalExceptions)
                 {
                     var diagnosticDescriptor = LinqDiagnosticsRules.DiagnosticRuleNotSupportedLinqExpression;
-                    var decoratedMessage = DecorateMessage(mqlResult.Exception.InnerException?.Message ?? "Unsupported LINQ expression", driverVersion, settings);
+                    var typesMapper = new TypesMapper(MqlGeneratorSyntaxElements.Linq.MqlGeneratorNamespace, context.TypesProcessor);
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqAnalyzer.cs:117]: Reuse the instance. And skip remapping if there is no exception message.

File: src/MongoDB.Analyzer/Core/TypesMapper.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+internal sealed class TypesMapper
+{
+    private const string RegexLookahead = "(?![\\w])";
+    private const string RegexLookbehind = "(?<![\\w\\.])";
+
+    private IDictionary<string, string> _typesRemapping;
+
+    public TypesMapper(string mqlGeneratorNamespace, TypesProcessor typesProcessor)
+    {
+        AddMappings(mqlGeneratorNamespace, typesProcessor);
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/TypesMapper.cs:28]: Lets do lazy initialization.

File: src/MongoDB.Analyzer/Core/TypesMapper.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public string RemapTypes(string expression)
+    {
+        if (_typesRemapping != null)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/TypesMapper.cs:33]: With lazy init no need for null check.

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/TypesMapper.cs (hunk 4 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/TypesMapper.cs]: Done

File: src/MongoDB.Analyzer/Core/Linq/LinqAnalyzer.cs (hunk 3 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqAnalyzer.cs]: Made it a shared instance among Builders/LINQ/POCO processing

**Comment by BorisDog** [comment]
Continued in [new PR](https://github.com/mongodb/mongo-csharp-analyzer/pull/47)

**Pull request closed without merging**